### PR TITLE
meta: support application fields (CRAFT-814)

### PR DESCRIPTION
**Pull request by mr-cal**
- [X] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [X] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [X] Have you successfully run `./runtests.sh static`?
- [X] Have you successfully run `./runtests.sh tests/unit`?

-----

Support for application fields in `snapcraft/7.0`.

### CODE REVIEW

File: snapcraft/meta/snap_yaml.py [hunk 1]
```diff
@@ -18,27 +18,58 @@
...
+    TODO: implement desktop (CRAFT-804)
+    TODO: implement extensions (CRAFT-805)
+    TODO: implement passthrough (CRAFT-854)
+    TODO: implement slots (CRAFT-816)
     """
 
     command: str
-    command_chain: List[str]
-    environment: Optional[Dict[str, Any]]
+    autostart: Optional[str]
```
Comment (mr-cal) [inline snapcraft/meta/snap_yaml.py:49]: Struggling with pylint:
```
************* Module snapcraft.commands.__init__
snapcraft/commands/__init__.py:1:0: R0801: Similar lines in 2 files
==snapcraft.meta.snap_yaml:[51:60]
==snapcraft.projects:[86:95]
    completer: Optional[str]
    stop_command: Optional[str]
    post_stop_command: Optional[str]
    start_timeout: Optional[str]
    stop_timeout: Optional[str]
    watchdog_timeout: Optional[str]
    reload_command: Optional[str]
    restart_delay: Optional[str]
    timer: Optional[str] (duplicate-code)
```

It's not duplicate code, yet I can't find a good way to disable the `duplicate-code` warning in just these files.  
`# pylint: disable=duplicate-code` isn't supported (https://github.com/PyCQA/pylint/issues/214).

Comment (mr-cal) [reply snapcraft/meta/snap_yaml.py:49]: Silenced with https://github.com/mr-cal/snapcraft/blob/097b7f04c8dc38d5e206f39c3b066f614b2671db/pyproject.toml#L48

File: snapcraft/projects.py [hunk 2]
```diff
@@ -47,23 +47,43 @@ class Config:  # pylint: disable=too-few-public-methods
 # see https://github.com/samuelcolvin/pydantic/issues/975#issuecomment-551147305
 # fmt: off
 if TYPE_CHECKING:
-    CommandChainStr = str
     UniqueStrList = List[str]
-    UniqueAliasList = List[str]
 else:
-    CommandChainStr = constr(regex=r"^[A-Za-z0-9/._#:$-]*$")
     UniqueStrList = conlist(str, unique_items=True)
-    UniqueAliasList = conlist(constr(regex=r"^[a-zA-Z0-9][-_.a-zA-Z0-9]*$"), unique_items=True)
```
Comment (mr-cal) [inline snapcraft/projects.py:56]: Catching errors with `@pydantic.validator` functions makes for more descriptive errors.

**Comment by sergiusens** [comment]
does not match -> is not a valid alias:

    testtools.matchers._impl.MismatchError: "The 'apps/test/aliases[0]' property does not match the required schema: '.test' does not match " not in "Issues while validating snapcraft.yaml: The 'apps/test/aliases[0]' property does not match the required schema: '.test' is not a valid alias. Aliases must be strings, begin with an ASCII alphanumeric character, and can only use ASCII alphanumeric characters and the following special characters: . _ -"

**Comment by mr-cal** [comment]
> does not match -> is not a valid alias:

Fixed.

**Review by sergiusens — APPROVED** [review]

**Pull request merged**